modules/c++/six.sicd/include/six/sicd/Utilities.h

```diff
@@ -19,30 +19,172 @@
  * see <http://www.gnu.org/licenses/>.
  *
  */
-#ifndef __SIX_SICD_UTILITIES_H__
-#define __SIX_SICD_UTILITIES_H__
+#ifndef __SIX_SICD_UTILITIES_H__
+#define __SIX_SICD_UTILITIES_H__
 
 #include <memory>
 #include <string>
 #include <vector>
+#include <utility>
 
 #include <scene/SceneGeometry.h>
 #include <scene/ProjectionModel.h>
-#include <six/sicd/ComplexData.h>
-
-namespace six
-{
-namespace sicd
-{
-class Utilities
-{
-public:
+#include <six/sicd/ComplexData.h>
+
+#include <six/NITFReadControl.h>
+
+namespace six
+{
+namespace sicd
+{
+class Utilities
+{
+public:
     static scene::SceneGeometry* getSceneGeometry(const ComplexData* data);
-
-    static scene::ProjectionModel* getProjectionModel(const ComplexData* data, 
+
+    static scene::ProjectionModel* getProjectionModel(const ComplexData* data, 
             const scene::SceneGeometry* geom);
 
     /*
+     * SicdContents is a helper object for readSicdContents
+     *
+     * This class is a simple wrapper around SICD information 
+     * designed to allow safe access to the multiple variables 
+     * returned by readSicdContents.
+     *
+     */
+    struct SicdContents {
+    
+        SicdContents()
+            : complexData(0),
+              widebandData()
+        {
+        }
+
+        SicdContents(const SicdContents& other)
+            : complexData(0),
+              widebandData(other.widebandData)
+        {
+            if(other.complexData.get())
+            {
+                complexData.reset(reinterpret_cast<ComplexData*>
+                                    (other.complexData->clone()));
+            }
+        }
+
+        SicdContents& operator=(const SicdContents& other)
+        {
+            if(other.complexData.get())
+            {
+                complexData.reset(reinterpret_cast<ComplexData*>
+                                    (other.complexData->clone()));
+                widebandData = other.widebandData;
+            }
+            return *this;
+        }
+
+        std::auto_ptr<ComplexData> complexData;
+        std::vector<std::complex<float> > widebandData;
+    };
+
+    /*
+     * Given a SICD path name and a list of schema, this function reads
+     * and parses the SICD in order to provide the wideband data as well
+     * as the ComplexData associated with the image..
+     *
+     * \param sicdPathname SICD NITF pathname
+     * \param schemaPaths One or more files or directories containing SICD 
+     * schemas
+     *
+     * \return ComplexData associated with the SICD NITF,
+     *           allocated with new
+     *         std::complex<float> buffer containing SICD wideband data,
+     *           allocated with new[]
+     *
+     * \throws See six::sicd::Utilities::getComplexData and 
+     *           six::sicd::Utilities::getWidebandData
+     *
+     */
+    static SicdContents readSicd(const std::string& sicdPathname,
```

Comment:
Rather than the complexity of a new `SicdContents` class (and the copy that will occur upon returning this unless this is a C++11 compiler or it's smart enough to do return value optimization), just have this function take in an auto_ptr to ComplexData by ref and a vector of complex float by ref.